### PR TITLE
Expand API

### PR DESCRIPTION
**Pull request by JeremyTCD**
- No changes to existing API
- ~~Add members with "params object[] args" for convenience.~~
  ~~E.g so instead of this:~~
    ```csharp
    InvokeFromStringAsync<int>(javascriptModule, args: new object[] { 3, 5 });
    ```
  ~~we can do this:~~
    ```csharp
    InvokeFromStringAsync<int>(javascriptModule, 3, 5);
    ```
- Add members with factory methods to atomically check if cached and cache if not cached.
    E.g:
    ```csharp
    Task<T> InvokeFromStringAsync<T>(Func<string> moduleFactory, string cacheIdentifier, string exportName, params object[] args);
    ```
    ```csharp
    Task<T> InvokeFromStreamAsync<T>(Func<Stream> moduleFactory, string cacheIdentifier, string exportName, params object[] args);
    ```
  so instead of this:
    ```csharp
    (bool success, Result result) = await nodeJSService.TryInvokeFromCacheAsync<Result>(cacheIdentifier, args: new[] { "success" });
    if(!success)
    {
        string moduleString = "module.exports = (callback, message) => callback(null, { resultMessage: message });"; 
        result = await nodeJSService.InvokeFromStringAsync<Result>(moduleString, cacheIdentifier, args: new[] { "success" });
    }
    ```
  we can do this:
    ```csharp
    Func<string> moduleFactory = () => "Retrieve from file, remote source etc";
    Result result = await nodeJSService.InvokeFromStringAsync<Result>(moduleFactory, cacheIdentifier, "success");
    ```
- Add members with no generic parameter for convenience when js doesn't return anything.
   E.g:
    ```csharp
    Task InvokeFromFileAsync(string modulePath, string exportName, CancellationToken cancellationToken, params object[] args);
    ```

Edit:  

No longer adding members with "params object[] args" because of the following ambiguity:
If we expose the following methods:
```csharp
InvokeFromStringAsync<string>(string moduleSource, string newCacheIdentifier, string exportName, params object[] args);
InvokeFromStringAsync<string>(string moduleSource, string newCacheIdentifier, params object[] args);
InvokeFromStringAsync<string>(string moduleSource, params object[] args);
```
Users might expect:
```csharp
InvokeFromStringAsync<string>(javascriptModule, arg1);
```
to invoke `InvokeFromStringAsync<string>(string moduleSource, params object[] args)`, but it actually invokes `InvokeFromStringAsync<string>(string moduleSource, string newCacheIdentifier, params object[] args);`.  

Similarly,
```csharp
InvokeFromStringAsync<string>(javascriptModule, arg1, arg2);
```
Invokes `InvokeFromStringAsync<string>(string moduleSource, string newCacheIdentifier, string exportName, params object[] args);`.

Some framework methods like [`Activator.CreateInstance`](https://docs.microsoft.com/en-us/dotnet/api/system.activator.createinstance?view=netcore-3.0#System_Activator_CreateInstance_System_Type_System_Object___) have the same issue. However,
with this library, I think there is higher probability of accidents because users might try `INodeServices.InvokeAsync` syntax like `InvokeFromFileAsync<string>(javascriptModule, arg1);` and it's likely that users will be sending strings to Node.js.

### CODE REVIEW

**Comment by DaniilSokolyuk** [comment]
@JeremyTCD 
LGTM! :)

**Comment by Taritsyn** [comment]
It would also be nice if the `GetFrom*` methods appeared, which would allow to work with non-function exports.

For example, now the following code leads to errors:

```csharp
using System;
using System.Threading.Tasks;

using Jering.Javascript.NodeJS;

namespace TestNodeJs
{
    class Program
    {
        static void Main(string[] args)
        {
            Task<double> task = StaticNodeJSService.InvokeFromStringAsync<double>(
                @"module.exports = { pi: 3.14 };",
                "math.js",
                "pi"
            );
            task.Wait();
            double result = task.Result;

            Console.WriteLine("result = {0}", result);
        }
    }
}
```

**Comment by JeremyTCD** [comment]
@DaniilSokolyuk Thanks for taking a look! Should be release by tomorrow.

@Taritsyn I'll keep that suggestion in mind, it's somewhat out of scope for now. Wrt #56, I realized we don't need new overloads, we just need `InvokeFrom*<object>` to return `ExpandoObject`s! Had a look at implementing it with this PR, but there were some blockers - System.Text.Json doesn't support `ExpandoObject`s yet. Also, to serialize a custom `undefined` in JSON we'll have to use a unique string, since JSON has no `undefined` type. This means custom JSON converters, which means  will have to be behind a flag so as not to affect performance of those who don't need `undefined`. Will handle this in a separate PR since there is quite a bit to do.

**Comment by Taritsyn** [comment]
@JeremyTCD OK.

File: src/NodeJS/NodeJSServiceImplementations/OutOfProcess/OutOfProcessNodeJSService.cs [hunk 1]
```diff
@@ -90,45 +90,114 @@ public abstract class OutOfProcessNodeJSService : INodeJSService
...
+
+            (bool success, T result) = await TryInvokeFromCacheAsync<T>(cacheIdentifier, exportName, args, cancellationToken).ConfigureAwait(false);
+
+            if (success)
+            {
+                return result;
+            }
+
+            // If module's exports doesn't exist in cache, or we aren't using the cache, create module and send it to the NodeJS process
+            var invocationRequest = new InvocationRequest(ModuleSourceType.Stream, null, cacheIdentifier, exportName, args, moduleFactory());
```
Comment (DaniilSokolyuk) [inline src/NodeJS/NodeJSServiceImplementations/OutOfProcess/OutOfProcessNodeJSService.cs]: Dispose stream please

**Comment by DaniilSokolyuk** [comment]
System.Text.Json is extandble
[Here](https://github.com/dotnet/runtime/blob/3cec4f50af9fc68bcde80952427092f7a8873864/src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs#L267) example how to deseriailize object like Newtonsoft.Json, we can add to JsonService as default converter

File: src/NodeJS/NodeJSServiceImplementations/OutOfProcess/OutOfProcessNodeJSService.cs (hunk 1 above)
Comment (JeremyTCD) [reply src/NodeJS/NodeJSServiceImplementations/OutOfProcess/OutOfProcessNodeJSService.cs]: Good catch

**Comment by JeremyTCD** [comment]
@DaniilSokolyuk 

Yeap it supports custom converters, but it doesn't support serializing to `ExpandoObject`s yet - 
 [JsonSerializer support for ExpandoObject](https://github.com/dotnet/corefx/issues/38007). So we can't implement this using `System.Text.Json` right now:

```csharp
dynamic result = nodeJSService.InvokeFromStringAsync<object>(moduleString);
```

**Comment by JeremyTCD** [comment]
🚀 Released v5.2.0.

**Pull request merged**